### PR TITLE
Filters posts powered by the API

### PR DESCRIPTION
**Pull request by chloealee**
#### What's this PR do?
- Changes Campaign Single Index page's initial state to be powered by API instead of sending signups from backend (`GET /posts`). 
- Powers filtering by status via the API (`GET /posts`).
- Adds relation to Post model of other related posts under the same signup (siblings).
- Updates Post Transformer to optionally include related posts (siblings) and signup object for `GET /posts`).
- Returns more information about the user in the User Transformer (user's last name, birthdate, email address, mobile number).  
- Returns the original photo url in the Post Transformer. 
- Updates `$include` param to allow for multiple values (e.g. `$include=signup,siblings`).
- Updates Post tests.  
- BONUS: 
    - Adds some missing things to README file for documentation.

#### How should this be reviewed?
- Test `campaigns/:nid` to make sure initial state looks as expected and filtering works properly. 
    - Make sure that all information about the signup and post is exposed correctly! 
    - Don't forget to check the "Original Photo" and "Edited Photo" links and that they're working correctly! 
- Make sure `/campaigns/:nid/inbox` isn't broken. 

#### Any background context you want to provide?
- I think that this breaks pagination? Is this ok for now and we can make another issue to fix this? 
- @DFurnes I don't think we need to do this but should we update the activity test to make sure it includes additional information about the user? Has same functionality so not sure if this is needed! 

#### Relevant tickets
Fixes https://www.pivotaltracker.com/n/projects/2019429/stories/149809538

#### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

### CODE REVIEW

**Comment by chloealee** [comment]
@DFurnes looks like the wercker tests that are failing have to do with the activity tests - maybe I'll rebase after you merge #347 to see if they're still failing and if so, will fix!

File: app/Http/Controllers/Api/ActivityController.php [hunk 1]
```diff
@@ -27,9 +27,22 @@ public function index(Request $request)
...
+        if (array_key_exists('status', $filters)) {
+            // Constrain Eager Load, only returning posts with status query.
+            $query = $query->with(['posts' => function ($query) use ($filters) {
+                $query->where('status', $filters['status']);
+            }]);
+
+            // Remove status from $filters to query Signups.
+            unset($filters['status']);
+        } else {
+            $query->with('posts');
```
Comment (DFurnes) [inline app/Http/Controllers/Api/ActivityController.php]: Make sure you save the updated query here: `$query = $query->with('posts')`

Comment (chloealee) [inline app/Http/Controllers/Api/ActivityController.php]: oh oops! thanks for catching that!

**Comment by chloealee** [comment]
@DFurnes looks like the tests are still failing :( When you have a chance, would you mind looking over with me?

File: app/Http/Controllers/Api/ActivityController.php [hunk 2]
```diff
@@ -27,9 +27,22 @@ public function index(Request $request)
...
+        $query = $this->newQuery(Signup::class);
 
         $filters = $request->query('filter');
+
+        if (array_has($filters, 'status')) {
+            // Constrain Eager Load, only returning posts with status query.
+            $query = $query->with(['posts' => function ($query) use ($filters) {
+                $query->where('status', $filters['status']);
+            }]);
+
```
Comment (chloealee) [inline app/Http/Controllers/Api/ActivityController.php]: @DFurnes we don't necessarily need this anymore since we are using the new `/posts` endpoint with siblings(!) but should I keep in case we need in the future? cc @sbsmith86

**Comment by chloealee** [comment]
@DFurnes when I only run the `PostModelTest`s, everything passes. However, when I run all of them, I get this error: 

```
There was 1 error:

1) PostModelTest::testSiblingRelationship
ErrorException: Trying to get property of non-object

/home/vagrant/DoSomething.org/rogue/tests/Models/PostModelTest.php:49
```
have you seen this before?

File: app/Http/Controllers/Api/ActivityController.php (hunk 2 above)
Comment (DFurnes) [inline app/Http/Controllers/Api/ActivityController.php]: Hmm, I don't have strong opinions but if we don't think we'll need it, it's always fun to delete code. 🔥

File: resources/assets/components/InboxItem/index.js [hunk 3]
```diff
@@ -22,8 +22,10 @@ class InboxItem extends React.Component {
     const post_id = post['id'];
     const signup = this.props.details.signup;
 
+    var posts = signup.posts ? signup.posts : post.siblings.data;
```
Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Hmm, this feels a little awkward with the implicit dependency on the `details` prop. I don't have any great suggestions off the top of my head, but perhaps something we can simplify a bit.

File: resources/assets/components/InboxItem/index.js [hunk 4]
```diff
@@ -39,6 +41,22 @@ class InboxItem extends React.Component {
     const campaign = this.props.details.campaign;
     const signup = this.props.details.signup;
 
+    if (post['user']) {
```
Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Hmm, this conditional bums me out a little. What are the two different scenarios that we get here?

File: resources/assets/components/InboxItem/index.js [hunk 5]
```diff
@@ -52,12 +70,12 @@ class InboxItem extends React.Component {
           </ul>
         </div>
         <div className="container__block -third">
-          {post['user'] ?
+          {post['user'] || signup.user ?
             <div>
-              <h2>{post['user']['first_name']} {post['user']['last_name']}, {calculateAge(post['user']['birthdate'])}</h2>
+              <h2>{first_name} {last_name}, {birthdate}</h2>
```
Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: It might make sense to use the Northstar user's [`displayName`](https://github.com/DoSomething/gateway/blob/30565f68e35fe1051be0569c8292e643f9a8c499/src/Resources/NorthstarUser.php#L21-L36) here!

File: resources/assets/components/CampaignSingle/index.js [hunk 6]
```diff
@@ -34,9 +23,48 @@ class CampaignSingle extends React.Component {
     this.filterPosts = this.filterPosts.bind(this);
   }
 
+  componentDidMount() {
+    this.api.get('api/v2/posts', {
+      filter: {
+        status: 'accepted',
+        campaign_id: this.props.campaign.id,
+      },
+      include: 'signup,siblings',
```
Comment (chloealee) [inline resources/assets/components/CampaignSingle/index.js]: You're right, it doesn't work yet :( But I agree - that looks a lot better! Should I great an issue for that in Gateway?

File: resources/assets/components/InboxItem/InboxTile.js [hunk 7]
```diff
@@ -8,7 +8,7 @@ class InboxTile extends React.Component {
 
     return (
       <li>
-        <img src={getImageUrlFromProp(post)}/>
+        <img src={getImageUrlFromProp(post) ? getImageUrlFromProp(post) : post.media['url']}/>
```
Comment (chloealee) [inline resources/assets/components/InboxItem/InboxTile.js]: ooo yes I like that a lot better!

File: resources/assets/components/CampaignSingle/index.js (hunk 6 above)
Comment (DFurnes) [inline resources/assets/components/CampaignSingle/index.js]: That's something we could probably fix in the `TransformsRequests` trait here!

Comment (DFurnes) [inline resources/assets/components/CampaignSingle/index.js]: Let's dig into this together sometime this week (and definitely doesn't block merging).

**Comment by sbsmith86** [comment]
> I think that this breaks pagination? Is this ok for now and we can make another issue to fix this?

We have an issue for this already. I pulled it into this sprint for myself to work on because I had some bandwidth but was waiting on some of your work to proceed. Since you are pulling in paginated posts we just need to get the pagination buttons hooked up. https://www.pivotaltracker.com/story/show/149809548

**Comment by chloealee** [comment]
ok awesome, thanks @sbsmith86 !

File: resources/assets/components/CampaignSingle/index.js (hunk 6 above)
Comment (chloealee) [inline resources/assets/components/CampaignSingle/index.js]: ok awesome!

File: resources/assets/components/InboxItem/index.js (hunk 4 above)
Comment (chloealee) [inline resources/assets/components/InboxItem/index.js]: @DFurnes the first scenario with `post['user']` is for populating the inbox items for the Campaign Inbox page. It looks like there is a TODO [here](https://github.com/DoSomething/rogue/blob/master/app/Http/Controllers/CampaignsController.php#L64-L75) suggesting to grab the user with the signup which would actually solve this problem and then we wouldn't have to do a conditional! Should I work on that or is that outside of the scope of this PR?

File: resources/assets/components/CampaignSingle/index.js [hunk 8]
```diff
@@ -34,9 +23,48 @@ class CampaignSingle extends React.Component {
...
+      postTotals: json.meta.pagination.total,
+      displayHistoryModal: null,
+      historyModalId: null,
+      nextPage: json.meta.pagination.links.next ? json.meta.pagination.links.next : null,
+      prevPage: json.meta.pagination.links.previous ? json.meta.pagination.links.previous : null,
+    }));
+  }
+
   // Filter posts based on status.
   filterPosts(status) {
```
Comment (chloealee) [inline resources/assets/components/CampaignSingle/index.js]: ah yes! updated :)

File: resources/assets/components/InboxItem/index.js (hunk 5 above)
Comment (chloealee) [inline resources/assets/components/InboxItem/index.js]: ohh I like this idea! Especially since we were talking about how it's a little weird to have a space between first name and comma if there is no last name. Would I have to make a call to gateway to get this?

File: resources/assets/components/InboxItem/index.js (hunk 3 above)
Comment (chloealee) [inline resources/assets/components/InboxItem/index.js]: @DFurnes would it be better to pass props like this: 
`key={key} post={post} campaign={campaign} signup={post.signup.data}` instead of `details={{post: post, campaign: campaign, signup: this.state.signups[post.signup_id]}}` [here](https://github.com/DoSomething/rogue/blob/master/resources/assets/components/CampaignSingle/index.js#L109)?

Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Yeah I think that's a great idea 👍 👍

File: resources/assets/components/InboxItem/index.js (hunk 4 above)
Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Got it! Nah, let's add a TODO referencing that & if you're feeling up to it we can always tackle in another PR afterwards.

File: resources/assets/components/InboxItem/index.js (hunk 5 above)
Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: It should already be available on the user response that's returned from Gateway in each case, but just not automatically bundled when using `toArray()`. So you should be able to attach it as another field in the response [here](https://github.com/DoSomething/rogue/blob/ac76c91a7607afd55adee9f523270ee3c3cc9561/app/Http/Controllers/CampaignsController.php#L75) and [here](https://github.com/DoSomething/rogue/blob/ac76c91a7607afd55adee9f523270ee3c3cc9561/app/Http/Controllers/CampaignsController.php#L108):

```diff
- $item->user = $user->toArray() ?: ['first_name' => 'A Doer'];
+ $item->user = [
+   'displayName' => $user->displayName(),
+   'birthdate' => $user->birthdate,
+   // other fields you want to include...
+ ];
```

Comment (chloealee) [inline resources/assets/components/InboxItem/index.js]: ahh I see - so cool! For the [second example](https://github.com/DoSomething/rogue/blob/ac76c91a7607afd55adee9f523270ee3c3cc9561/app/Http/Controllers/CampaignsController.php#L108), we're not actually passing `$signups` anymore since this PR grabs it via API call. Is it ok to do for the first (in the `showInbox` function and then when rendering the user information, doing a conditional like `if (displayName)` use that info and if not get the user info another way?

Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Oh right! For the signup API call, you can add the `$user->displayName` to [the transformer](https://github.com/DoSomething/rogue/blob/master/app/Http/Transformers/UserTransformer.php).

Comment (chloealee) [inline resources/assets/components/InboxItem/index.js]: when I tried to do the above for the Campaign Inbox page (`showInbox` function) and logged out the user, I got 
```
Object
  birthdate: "1989-04-12"
  displayName: "5978afc27f43c27d8755558c"
```
which seemed to grab the user ID instead of display name?

Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: And actually it might be nice to use the transformer here too so that we only deal with formatting user responses in one place:

```
$item->user = new UserTransformer($user);
```

Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Ahhhhhhhh yeah if we don't have a last name for the user [we return their ID](https://github.com/DoSomething/gateway/blob/30565f68e35fe1051be0569c8292e643f9a8c499/src/Resources/NorthstarUser.php#L35). That's maybe something to update in Gateway. 😖

Comment (DFurnes) [inline resources/assets/components/InboxItem/index.js]: Let's break this out into another PR since it's getting more involved!

File: resources/assets/components/CampaignSingle/index.js [hunk 9]
```diff
@@ -91,27 +84,44 @@ class CampaignSingle extends React.Component {
     this.hideHistory();
   }
 
+  // Make API call to GET /posts to get posts by filtered status.
+  getPostsByStatus(status, campaignId) {
+  this.api = new RestApiClient;
```
Comment (DFurnes) [inline resources/assets/components/CampaignSingle/index.js]: Sorry one last little thing: looks like this method isn't indented! 👾

File: resources/assets/components/InboxItem/index.js (hunk 5 above)
Comment (chloealee) [inline resources/assets/components/InboxItem/index.js]: ok sounds good! I will make a tiny card for this :)

File: resources/assets/components/CampaignSingle/index.js (hunk 9 above)
Comment (chloealee) [inline resources/assets/components/CampaignSingle/index.js]: oh oops - thank you! Updated :)

**Comment by chloealee** [comment]
oof @DFurnes still getting [this error](https://github.com/DoSomething/rogue/pull/348#issuecomment-321347297) as before :-/

**Comment by DFurnes** [comment]
That's frustrating! 😦 Can you try dropping a `dd()` above that line to see what's going wrong?

**Comment by chloealee** [comment]
I'm getting a new error now than I was yesterday! It doesn't even seem to get to the Post Model Test: 
```
PHP Fatal error:  Class 'FakerNorthstarId' not found in /home/vagrant/DoSomething.org/rogue/database/factories/ModelFactory.php on line 47
```
I tried to include that at the top of `ModelFactory.php` but that didn't seem to do anything...

**Comment by DFurnes** [comment]
@chloealee Try running `composer dump-autoload`!

**Comment by chloealee** [comment]
@DFurnes thank you! I updated the tests so it should be working now! The problem was that we assumed it the Signups that were being created were with `id`s 1-5 but they were actually with `id`s in the upper 70s. Just changed this so we don't assume!
